heartbeat/heartbeat.go

```diff
@@ -0,0 +1,214 @@
+	electionName   string
+	tls            config.TLS
+	choria         config.ChoriaConnection
+	leaderElection bool
+	interval       time.Duration
+	headers        map[string]string
+	subjects       []*Subject
+	log            *logrus.Entry
+	paused         atomic.Bool
+}
+
+type Subject struct {
+	name     string
+	interval time.Duration
+	headers  map[string]string
+}
+
+// New creates a new instance of the Heartbeat struct
+func New(hbcfg *config.HeartBeat, replicatorName string, log *logrus.Entry) (*HeartBeat, error) {
+	hb := &HeartBeat{
+		replicatorName: replicatorName,
+		electionName:   fmt.Sprintf("%s_HB", replicatorName),
+		tls:            hbcfg.TLS,
+		choria:         hbcfg.Choria,
+		leaderElection: hbcfg.LeaderElection,
+		headers:        hbcfg.Headers,
+		log:            log,
+		url:            hbcfg.URL,
+	}
+
+	hb.paused.Store(hb.leaderElection)
+
+	intervalString := _DEFAULTUPDATEINTERVAL_
+	if hbcfg.Interval != "" {
+		intervalString = hbcfg.Interval
+	}
+	dur, err := time.ParseDuration(intervalString)
+	if err != nil {
+		return nil, err
+	}
+	hb.interval = dur
+
+	for _, s := range hbcfg.Subjects {
+		sub := &Subject{name: s.Name}
+		if s.Interval == "" {
+			sub.interval = hb.interval
+		} else {
+			sdur, err := time.ParseDuration(s.Interval)
+			if err != nil {
+				return nil, err
+			}
+			sub.interval = sdur
+		}
+		mergedHeaders := hb.headers
+		for k, v := range s.Headers {
+			mergedHeaders[k] = v
+		}
+		sub.headers = mergedHeaders
+		hb.subjects = append(hb.subjects, sub)
+	}
+
+	return hb, nil
+}
+
+// Run initializes a the jetstream connection and spawns a go routine for every configured subject
+// that will publish a heartbeat message on the defined interval
+func (hb *HeartBeat) Run(ctx context.Context, wg *sync.WaitGroup) error {
+	nc, err := util.ConnectNats(ctx, "subject-heartbeats", hb.url, &hb.tls, &hb.choria, false, hb.log)
+	if err != nil {
+		return err
+	}
+
+	if hb.leaderElection {
+		err = hb.setupElection(hb.replicatorName, nc, ctx)
+		if err != nil {
+			hb.log.Errorf("Could not set up elections: %v", err)
+			return err
+		}
+	}
+
+	for _, subject := range hb.subjects {
+		wg.Add(1)
+		hbSubjects.WithLabelValues(hb.replicatorName, subject.name).Inc()
+		go heartBeatWorker(subject, nc, hb.replicatorName, ctx, wg, &hb.paused, hb.log.WithField("subject", subject.name))
+	}
+
+	return nil
+}
+
+func heartBeatWorker(sub *Subject, nc *nats.Conn, replicatorName string, ctx context.Context, wg *sync.WaitGroup, paused *atomic.Bool, log *logrus.Entry) {
+	defer wg.Done()
+	wasPaused := false
+	js, err := nc.JetStream()
+	if err != nil {
+		log.Warnf("Unable to create jetstream connection: %v", err)
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		log.Warn("Unable to determine hostname. Publishing heartbeats with empty 'originator' header.")
+	}
+
+	msg := nats.NewMsg(sub.name)
+	for k, v := range sub.headers {
+		msg.Header.Add(k, v)
+	}
+	msg.Header.Add("originator", hostname)
```

Comment:
add `Choria-SR-Subject` header to contain the subject being published to, duplicating the info in msg.Subject